contrail_api_cli/commands/batch.py

```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+import shlex
+import sys
+import fileinput
+from six import text_type
+
+from keystoneauth1.exceptions.http import HttpError, HTTPClientError
+
+from ..command import Command, Arg, Option
+from ..manager import CommandManager
+from ..context import Context
+from ..schema import SchemaError
+from ..exceptions import CommandError, CommandNotFound, NotFound, Exists
+from ..utils import printo
+
+
+class Batch(Command):
+    """ Execute commands from file(s) or from stdin
+    """
+    description = "Run commands from a batch file/stdin"
+    files = Arg(nargs="*", help="Name of the batch file")
+
+    def __call__(self, files=None):
+        manager = CommandManager()
+        try:
+            for line in fileinput.input(files=files):
+                if line[0] == '#':
+                    continue
+                action = shlex.split(line.rstrip())
+                if len(action) < 1:
+                    continue
+                cmd = manager.get(action[0])
+                args = action[1:]
+                result = cmd.parse_and_call(*args)
+                if result:
+                    printo(result)
+        except IOError:
+            printo("Cannot read from file: {}".format(filename))
```

Comment:
This is by intention: if wrong filename is given, I do not want the whole `contrail-api-cli` to stop with a Python traceback, just print the error, and gracefully exit if run from cli, or remain in `contrail-api-cli` shell, if batch command is executed from the shell.

Throwing the error and exiting with traceback can also make sense if someone would like to run `contrail-api-cli batch ...` from a shell script, and he would like to check the exit code.

Reply:
It will not throw a traceback, main.py catches CommandError exceptions, prints the error message to stderr and exit with 1

See: https://github.com/eonpatapon/contrail-api-cli/blob/master/contrail_api_cli/main.py#L84

Reply:
Yes, you are right again. I mixed this with the case when wrong command is given. The `CommandNotFound` is not handled by the `shell` command code when it runs a command. Of course, because it is doing the command lookup before, and there it handles the `CommandNotFound`.
I enhanced the error handling, and changed to raise `CommandError` in case of file IO error, or when the command lookup fails in the `batch.py` code.